### PR TITLE
[SPARK-34094][SQL] Extends StringTranslate to support unicode characters whose code point >= U+10000

### PR DESCRIPTION
**Pull request by sarutak**
### What changes were proposed in this pull request?

This PR extends `StringTranslate` to support unicode characters whose code point >= `U+10000`.

### Why are the changes needed?

To make it work with wide variety of characters.

### Does this PR introduce _any_ user-facing change?

Yes. Users can use `StringTranslate` with unicode characters whose code point >= `U+10000`.

### How was this patch tested?

New assertion added to the existing test.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
cc @srowen FYI

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 1]
```diff
@@ -1065,16 +1065,19 @@ public UTF8String replace(UTF8String search, UTF8String replace) {
...
     String srcStr = this.toString();
 
     StringBuilder sb = new StringBuilder();
-    for(int k = 0; k< srcStr.length(); k++) {
-      if (null == dict.get(srcStr.charAt(k))) {
-        sb.append(srcStr.charAt(k));
-      } else if ('\0' != dict.get(srcStr.charAt(k))){
-        sb.append(dict.get(srcStr.charAt(k)));
+    int charCount = 0;
+    for(int k = 0; k < srcStr.length(); k += charCount) {
```
Comment (srowen) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java:1073]: Nit: add a space after for

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 2]
```diff
@@ -1065,16 +1065,19 @@ public UTF8String replace(UTF8String search, UTF8String replace) {
...
+      int codePoint = srcStr.codePointAt(k);
+      charCount = Character.charCount(codePoint);
+      String subStr = srcStr.substring(k, k + charCount);
+      if (null == dict.get(subStr)) {
+        sb.append(subStr);
+      } else if (!"\0".equals(dict.get(subStr))) {
```
Comment (srowen) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java:1079]: While we're here, we could probably optimize away the two calls to dict.get(subStr)

**Review by srowen — CHANGES_REQUESTED** [review]

File: common/unsafe/src/test/java/org/apache/spark/unsafe/types/UTF8StringSuite.java [hunk 3]
```diff
@@ -465,27 +465,27 @@ public void translate() {
     assertEquals(
       fromString("1a2s3ae"),
       fromString("translate").translate(ImmutableMap.of(
-        'r', '1',
-        'n', '2',
-        'l', '3',
-        't', '\0'
+        "r", "1",
```
Comment (srowen) [inline common/unsafe/src/test/java/org/apache/spark/unsafe/types/UTF8StringSuite.java:468]: Why do these need to be strings now? all seem to be 1 char.

Comment (sarutak) [reply common/unsafe/src/test/java/org/apache/spark/unsafe/types/UTF8StringSuite.java:468]: Ah, it's a wrong change.

Comment (sarutak) [reply common/unsafe/src/test/java/org/apache/spark/unsafe/types/UTF8StringSuite.java:468]: The type parameter of `translate` is changed from `Map<Character, Character>` to `Map<String, String>` to support >=`U+10000` unicode characters so this change is necessary. Those characters can't be represent as `Character`.

**Comment by srowen** [comment]
OK so it doesn't show up in the tests, but these code points would be more than one char? is there any easy way to add a test for this case then? that seems to be the purpose of this change.

**Comment by sarutak** [comment]
> OK so it doesn't show up in the tests, but these code points would be more than one char?

@srowen 
It's already added to `StringExpressionSuite`. For example,  `\uD840\uDC0B` [here](https://github.com/apache/spark/pull/31164/files#diff-fb37dc07c90c672fb74f787b8881f0e335386d69b49c9eed065b20f4a09d5c9fR556) is one Unicode character which is needed to be represented as two `char` values.

**Review by srowen — APPROVED** [review]

**Comment by srowen** [comment]
Merged to master

**Pull request closed without merging**